cmd/registry/controller/controller.go

```diff
@@ -144,107 +136,249 @@ func generateActions(
+	parentList, err := ListResources(ctx, client, parentName, "")
+	if err != nil {
+		return nil, err
+	}
 
-			if takeAction {
-				// Since the GeneratedResource is non-existent here,
-				// we will have to derive the exact name of the target resourc
-				resourceName, err := resourceNameFromGroupKey(
-					resourcePattern, groupKey)
-				if err != nil {
-					log.Debugf(ctx, "skipping entry for %q, cannot derive target resource name for pattern: %q", groupKey, resourcePattern)
-					continue
-				}
-				cmd, err := generateCommand(generatedResource.Action, resourceName)
-				if err != nil {
-					return nil, err
-				}
-
-				action := &Action{
-					Command:           cmd,
-					GeneratedResource: resourceName,
-					RequiresReceipt:   generatedResource.Receipt,
-				}
-				actions = append(actions, action)
-			}
+	for _, parent := range parentList {
+		// Skip if this parent was already visited.
+		if visited[parent.GetResourceName().String()] {
+			continue
+		}
+		// Since the GeneratedResource is non-existent here,
+		// we will have to derive the exact name of the target resource
+		targetResourceName, err := resourceNameFromParent(resourcePattern, parent.GetResourceName().String())
+		if err != nil {
+			log.FromContext(ctx).WithError(err).Error("Cannot generate target resourceName to be created.")
+			continue
+		}
+
+		takeAction, err := needsCreate(
+			targetResourceName,
+			dependencyMaps,
+			generatedResource,
+		)
+
+		if err != nil {
+			log.Errorf(ctx, "%s", err)
+			continue
+		}
 
+		if takeAction {
+			cmd, err := generateCommand(generatedResource.Action, targetResourceName.String())
+			if err != nil {
+				return nil, fmt.Errorf("Cannot generate command: %s", err)
+			}
+			a := &Action{
+				Command:           cmd,
+				GeneratedResource: targetResourceName.String(),
+				RequiresReceipt:   generatedResource.Receipt,
+			}
+			actions = append(actions, a)
 		}
 	}
 
 	return actions, nil
 }
+
+func needsUpdate(
+	targetResourceName ResourceName,
+	targetResourceTime time.Time,
+	dependencyMaps []map[string]time.Time,
+	generatedResource *rpc.GeneratedResource,
+	createMode bool) (bool, error) {
+	takeAction := false
+	//Check if all the dependencies exist in the map.
+	for i, dependency := range generatedResource.Dependencies {
+		dMap := dependencyMaps[i]
+		// Get the entity to look for in dependencyMap
+		entityKey, err := getEntityKey(dependency.Pattern, targetResourceName)
+		if err != nil {
+			// This means that there is error in the pattern definition, hence return
+			return false, fmt.Errorf("cannot match resource with dependency. Error: %s", err.Error())
+		}
+
+		// All the dependencies should be present to generate an action.
+		if maxUpdateTime, ok := dMap[entityKey]; ok {
+			if maxUpdateTime.After(targetResourceTime) {
+				takeAction = true // Take action if dependency timestamp is later than resource timestamp
+			}
+		} else {
+			takeAction = false
+			break
```

Comment:
This should `return false` directly for clarity.

Reply:
Done.